constantine/serialization/codecs_ecdsa_secp256k1.nim

```diff
@@ -0,0 +1,163 @@
+  res.rawCopy(39, Secp256k1Oid, 0, 9)
+
+proc toPemPublicKey*(res: var array[88, byte], publicKey: EC_ShortW_Aff[Fp[C], G1]) =
+  ## Encodes a public key as ASN.1 DER encoded public keys.
+  ##
+  ## See: https://www.secg.org/sec1-v2.pdf appendix C.3
+  ##
+  ## TODO: Adjust to support different curves.
+  # Start with SEQUENCE
+  res.rawCopy(0, [byte(0x30), byte(0x56)], 0, 2)
+
+  # Algorithm identifier
+  const algoId = [
+    byte(0x30), byte(0x10),                    # SEQUENCE
+    byte(0x06), byte(0x07),                    # OID for EC
+    byte(0x2A), byte(0x86), byte(0x48),        # 1.2.840.10045.2.1
+    byte(0xCE), byte(0x3D), byte(0x02), byte(0x01),
+    byte(0x06), byte(0x05),                    # OID for secp256k1
+    byte(0x2B), byte(0x81), byte(0x04), byte(0x00), byte(0x0A) # 1.3.132.0.10
+  ]
+
+  res.rawCopy(2, algoId, 0, algoId.len) # algoId.len == 18
+
+  # Public key as bit string
+  const encoding = [byte(0x03), byte(0x42)] # [BIT-STRING, 2+32+32 prefix & coordinates]
+  const prefix = [
+    byte(0x00),  # DER BIT STRING: number of unused bits (always 0 for keys)
+    byte(0x04)   # SEC1: uncompressed point format marker
+  ]
+
+  template toByteArray(x: Fp[C] | Fr[C]): untyped =
+    var a: array[32, byte]
+    a.toBytes(x)
+    a
+
+  res.rawCopy(20, encoding, 0, 2)
+  res.rawCopy(22, prefix, 0, 2)
+  res.rawCopy(24, publicKey.x.toByteArray(), 0, 32)
+  res.rawCopy(56, publicKey.y.toByteArray(), 0, 32)
+
+## NOTE:
+## The below procs / code is currently "unsuited" for Constantine in the sense that
+## it currently still contains stdlib dependencies. Most of those are trivial, with the
+## exception of a base64 encoder.
+## Having a ANS1.DER encoder (and maybe decoder in the future) for SEC1 private and
+## public keys would be nice to have in CTT, I think (at least for the curves that
+## we support for the related operations; secp256k1 at the moment).
+
+proc wrap(s: string, maxLineWidth = 64): string =
+  ## Wrap the given string at `maxLineWidth` over multiple lines
+  let lines = s.len.ceilDiv maxLineWidth
+  result = newStringOfCap(s.len + lines)
+  for i in 0 ..< lines:
+    let frm = i * maxLineWidth
+    let to = min(s.len, (i+1) * maxLineWidth)
+    result.add s[frm ..< to]
+    if i < lines-1:
+      result.add "\n"
+
+proc toPemFile*(publicKey: PublicKey): string =
+  ## Convert a given private key to data in PEM format following SEC1
+  ##
+  ## RFC 7468 describes the textual encoding of these files:
+  ## https://www.rfc-editor.org/rfc/rfc7468#section-10
+  # 1. Convert public key to ASN.1 DER
+  var derB: array[88, byte]
+  privateAccess(PublicKey) # for `raw` access
+  derB.toPemPublicKey(publicKey.raw)
+  # 2. Encode bytes in base64
+  let der64 = derB.encode().wrap()
+  # 3. Wrap in begin/end public key template
+  result = "-----BEGIN PUBLIC KEY-----\n"
+  result.add der64 & "\n"
+  result.add "-----END PUBLIC KEY-----\n"
```

Comment:
It seems like the length of the result is fixed?
So in the refactoring issue for ECDSA serialization, we should replace the string output with a `var array` or `var openarray` (+ size check)